cpp/src/dbscan/dbscan.cuh

```diff
@@ -21,45 +21,74 @@
 
-// Default max mem set to a reasonable value for a 16gb card.
-template <typename T, typename Index_ = int>
-Index_ computeBatchCount(Index_ n_rows, size_t max_mbytes_per_batch) {
-  Index_ n_batches = 1;
-  // There seems to be a weird overflow bug with cutlass gemm kernels
-  // hence, artifically limiting to a smaller batchsize!
-  ///TODO: in future, when we bump up the underlying cutlass version, this should go away
-  // paving way to cudaMemGetInfo based workspace allocation
-
-  if (max_mbytes_per_batch <= 0) max_mbytes_per_batch = DEFAULT_MAX_MEM_MBYTES;
-
-  Index_ MAX_LABEL = std::numeric_limits<Index_>::max();
-
-  while (true) {
-    size_t batchSize = ceildiv<size_t>(n_rows, n_batches);
-    if (((batchSize * n_rows * sizeof(T) * 1e-6 < max_mbytes_per_batch) &&
-         /**
-          * Though single precision can be faster per execution of each kernel,
-          * there's a trade-off to be made between using single precision with
-          * many more batches (which become smaller as n_rows grows) and using
-          * double precision, which will have less batches but could become 8-10x
-          * slower per batch.
-          */
-         (batchSize * n_rows < MAX_LABEL)) ||
-        batchSize == 1)
-      break;
-    ++n_batches;
+template <typename Index_ = int>
+Index_ computeBatchCount(size_t &estimated_memory, Index_ n_rows,
+                         size_t max_mbytes_per_batch = 0,
+                         Index_ neigh_per_row = 0) {
+  // In real applications, it's unlikely that the sparse adjacency matrix
+  // comes even close to the worst-case memory usage, because if epsilon
+  // is so large that all points are connected to 10% or even more of other
+  // points, the clusters would probably not be interesting/relevant anymore
+  ///@todo: expose `neigh_per_row` to the user
+
+  if (neigh_per_row <= 0) neigh_per_row = n_rows;
+
+  // we'll estimate the memory consumption per row.
+  // First the dense adjacency matrix
+  estimated_memory = n_rows * sizeof(bool);
+  // sparse adjacency matrix
+  estimated_memory += neigh_per_row * sizeof(Index_);
+  // core points and two indicator variables
+  estimated_memory += 3 * sizeof(bool);
+  // the rest will be so small that it should fit into what we have left over
+  // from the over-estimation of the sparse adjacency matrix
+  estimated_memory *= n_rows;
+
+  if (max_mbytes_per_batch <= 0) {
+    /* using default here as in decision tree, waiting for mem info from device allocator
+    size_t total_mem;
+	  CUDA_CHECK(cudaMemGetInfo(&max_mbytes_per_batch, &total_mem));
+    */
+    max_mbytes_per_batch = DEFAULT_MAX_MEM_MBYTES;
+  }
+
+  Index_ nBatches =
+    (Index_)ceildiv<size_t>(estimated_memory, max_mbytes_per_batch * 1000000);
+  size_t MAX_LABEL = (size_t)std::numeric_limits<Index_>::max();
+  // n_rows * n_rows_per_batch < MAX_LABEL => n_rows * (n_rows / nBatches) < MAX_LABEL
+  // => nBatches >= n_rows * n_rows / MAX_LABEL
+  Index_ nBatchesPrec =
```

Comment:
Not sure I got the reasoning behind using `MAX_LABEL` for computing this batch count. Can you elaborate?

Reply:
This is still actually the same as before. Line 52 (for now) of `dbscan.cuh` has this exact test:
`batchSize * n_rows < MAX_LABEL`
And line 107 (for now) of `runner.cuh` also tests for this : `N * batchSize < MAX_LABEL`

For now, I haven't changed the indexing/addressing used in the actual algorithms because there doesn't seem to be a benefit after all (see also the comment above). So we still have to limit the batch-size according to `MAX_LABEL` even if we have more than 2GB of memory available and we estimate to use more than 2GB per batch.

Reply:
makes sense now. Thanks.